File: python/2dArrays/rotating.py
import copy
import math

# n = # rows = # columns in the given 2d array
def rotate(given_array, n):
  rotated = copy.deepcopy(given_array)
  for i in range(n):
    for j in range(n):
      (new_i, new_j) = rotate_sub(i, j, n)
      rotated[new_i][new_j] = given_array[i][j]
  return rotated

def rotate_in_place(given_array, n):
  for i in range(math.ceil(n/2)):
    for j in range(math.floor(n/2)):
      tmp = [-1] * 4
      (current_i, current_j) = (i,j)
      for k in range(4):
        tmp[k] = given_array[current_i][current_j]
        (current_i, current_j) = rotate_sub(current_i, current_j, n)
      for k in range(4):
        given_array[current_i][current_j] = tmp[(k - 1) % 4]
        (current_i, current_j) = rotate_sub(current_i, current_j, n)
  return given_array

def rotate_sub(i, j, n):
  return j, n - 1 - i

def to_string(given_array):
    list_rows = []
    for row in given_array:
        list_rows.append(str(row))
    return '[' + ',\n '.join(list_rows) + ']'

# NOTE: The following input values will be used for testing your solution.
a1 = [[1, 2, 3],
      [4, 5, 6],
      [7, 8, 9]]
print(rotate(a1, 3)) # should return:
# [[7, 4, 1],
#  [8, 5, 2],
#  [9, 6, 3]]
print(rotate_in_place(a1, 3)) # should return same

a2 = [[1, 2, 3, 4],
      [5, 6, 7, 8],
      [9, 10, 11, 12],
      [13, 14, 15, 16]]
print(rotate(a2, 4)) # should return:
# [[13, 9, 5, 1],
#  [14, 10, 6, 2],
#  [15, 11, 7, 3],
#  [16, 12, 8, 4]]
print(rotate_in_place(a2, 4)) # should return same

File: java/main/src/main/java/twoDmArray/Rotating.java
package twoDmArray;

import java.util.Arrays;

public class Rotating {
    public static void main(String[] args) {
        // NOTE: The following input values will be used for testing your solution.
        int a1[][] = {{1, 2, 3},
                {4, 5, 6},
                {7, 8, 9}};
        System.out.println(Arrays.deepToString(rotate(a1, 3)) + "\n");
        System.out.println(Arrays.deepToString(rotateInPlace(a1,3))+ "\n");
        // should return:
        // [[7, 4, 1],
        //  [8, 5, 2],
        //  [9, 6, 3]]

        int a2[][] = {{1, 2, 3, 4},
                {5, 6, 7, 8},
                {9, 10, 11, 12},
                {13, 14, 15, 16}};
        System.out.println(Arrays.deepToString(rotate(a2, 4)) + "\n");
        System.out.println(Arrays.deepToString(rotateInPlace(a2,4))+ "\n");
        // rotate(a2, 4) should return:
        // [[13, 9, 5, 1],
        //  [14, 10, 6, 2],
        //  [15, 11, 7, 3],
        //  [16, 12, 8, 4]]
    }

    public static int[][] rotate(int[][] a, int n) {
        int[][] rotated = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                rotated[j][n - 1 - i] = a[i][j];
            }
        }
        return rotated;
    }

    public static int[][] rotateInPlace(int[][] a, int n) {
        for (int i = 0; i < n / 2 + n % 2; i++) {
            for (int j = 0; j < n / 2; j++) {
                int[] tmp = new int[4];
                int currentI = i;
                int currentJ = j;
                for (int k = 0; k < 4; k++) {
                    tmp[k] = a[currentI][currentJ];
                    int[] newCoordinates = rotateSub(currentI, currentJ, n);
                    currentI = newCoordinates[0];
                    currentJ = newCoordinates[1];
                }
                for (int k = 0; k < 4; k++) {
                    a[currentI][currentJ] = tmp[(k + 3) % 4];
                    int[] newCoordinates = rotateSub(currentI, currentJ, n);
                    currentI = newCoordinates[0];
                    currentJ = newCoordinates[1];
                }
            }
        }
        return a;
    }

    private static int[] rotateSub(int i, int j, int n) {
        int[] newCoordinates = new int[2];
        newCoordinates[0] = j;
        newCoordinates[1] = n - 1 - i;
        return newCoordinates;
    }
}
